litho-core/src/main/java/com/facebook/litho/ComponentTree.java

```diff
@@ -1902,54 +1922,267 @@ protected LayoutState calculateLayoutState(
+        @Nullable final String extraAttribution) {
+      super(new Callable<LayoutState>() {
+        @Override
+        public LayoutState call() throws Exception {
+
+          final ComponentContext contextWithStateHandler;
+
+          synchronized (ComponentTree.this) {
+            final KeyHandler keyHandler = (ComponentsConfiguration.useGlobalKeys
+                || ComponentsConfiguration.isDebugModeEnabled)
+                    ? new KeyHandler(context.getLogger())
+                    : null;
+
+            contextWithStateHandler =
+                new ComponentContext(
+                    context, StateHandler.acquireNewInstance(mStateHandler), keyHandler, treeProps);
+          }
+
+          if (lock != null) {
+            synchronized (lock) {
+              return LayoutState.calculate(
+                  contextWithStateHandler,
+                  root,
+                  mId,
+                  widthSpec,
+                  heightSpec,
+                  diffingEnabled,
+                  previousLayoutState,
+                  mCanPrefetchDisplayLists,
+                  mCanCacheDrawingDisplayLists,
+                  mShouldClipChildren,
+                  mPersistInternalNodeTree,
+                  source,
+                  extraAttribution);
+            }
+          } else {
+            return LayoutState.calculate(
+                contextWithStateHandler,
+                root,
+                mId,
+                widthSpec,
+                heightSpec,
+                diffingEnabled,
+                previousLayoutState,
+                mCanPrefetchDisplayLists,
+                mCanCacheDrawingDisplayLists,
+                mShouldClipChildren,
+                mPersistInternalNodeTree,
+                source,
+                extraAttribution);
+          }
+        }
+      });
+
+      this.context = context;
+      this.root = root;
+      this.widthSpec = widthSpec;
+      this.heightSpec = heightSpec;
+      this.diffingEnabled = diffingEnabled;
+      this.previousLayoutState = previousLayoutState;
+      this.treeProps = treeProps;
+      this.extraAttribution = extraAttribution;
+    }
+
+    boolean isReleased() {
+      return isReleased;
+    }
+
+    @Override
+    public void run() {
+      if (runningThreadId.compareAndSet(-1, Process.myTid())) {
+        super.run();
+      }
+    }
+
+    @Override
+    public LayoutState get() {
+      final int runningThreadId = this.runningThreadId.get();
+      final int originalThreadPriority;
+      if (isMainThread() && !isDone() && runningThreadId != Process.myTid()) {
+        // Main thread is about to be blocked, raise the running thread priority.
+        originalThreadPriority = Process.getThreadPriority(runningThreadId);
+
+        boolean success = false;
+        int targetThreadPriority = Process.THREAD_PRIORITY_DISPLAY;
+        while (!success && targetThreadPriority < originalThreadPriority) {
```

Comment:
This is awesome, I love what we're doing here 👌 IMO, we could extract this priority increase code to a separate utility for cleanliness and because I think we might re-use it.

Reply:
This code could probably be replaced with a slightly refactored https://github.com/facebook/litho/blob/master/litho-core/src/main/java/com/facebook/litho/LayoutThreadFactory.java

Reply:
Sure, done.